### ISSUE TITLE
Hidden statement inputs seem broken

### ISSUE THREAD

**Issue by Marclev78** [google/blockly#1967]
### Problem
If I attach a block to a hidden statement input (reasonably expecting that block to be hidden), the block is programmatically attached to the input, but shown free-floating on the workspace.

### Expected behavior
The "visible" flag of the input should cascade down to anything attached to it.

### Actual Behavior

Actual behavior is basically "Weird".

I can verify programmatically that the block is attached to the input, but visually it just appears near the block with the input on the work-space and I drag it around free-form.

This makes hiding a statement input somewhat futile.

### Steps to reproduce
I have code like this:

```
        const summaryInput = this.appendStatementInput("summaryCollection");
        summaryInput.setVisible(false);
        summaryInput.connection.connect(blockToAttach.previousConnection);
```

blockToAttach ends up showing detached on the workspace (even though it is in fact attached in the block model).

### Workaround

As I can't see a way of hiding a block (looking at the Block API), I'm not sure what the workaround is.

**Comment by RoboErikG** [comment]
setVisible() is used when collapsing a block stack and we assumed devs wouldn't be modifying a user's blocks while they were hidden. As a workaround, you can toggle setVisible again or make a function to run the code that [typically hides a connected block](https://github.com/google/blockly/blob/master/core/input.js#L182) manually before attaching it.

As an aside, what's your use case?

**Comment by Marclev78** [comment]
I have some "advanced" stuff that I don't want to clutter up the main workspace with and only want to appear when the user opens the mutation dialog, so that they can configure it there.
As it happens, setCollapsed() better suits my use case as that gives some visual indication that there is more to the block than it seems.

Mabe the documentation for `setVisible` should be updated to reflect that the dev may have to call `hideAll` / `unhideAll` on the connection as well for it to work?

**Comment by RoboErikG** [comment]
Ah, setCollapsed() is probably the least work code wise, but your use case would work better with a custom collapse/uncollapse icon that mutates the block to show the hidden state. Anything can trigger a block mutation (it doesn't have to use the mutator dialog), so you could add a field_image with your collapse/uncollapse icon and mutate the block whenever it's clicked. We'll add a note to the js docs for that API.

Note: You do have to set the image field to editable to get click events on it (#1887 )

**Comment by RoboErikG** [comment]
This turns out to be more complicated than expected. There are quite a few cases that go through different paths and have differing state depending on the timing. #2003 is a start at a fix but needs a lot of tests to be sure all cases are covered correctly.

Notes from working on this:

Calling connect directly didn't create an undo group. Anyone else using connect as an API would have had to create an event group before calling it or they'd get into a weird state.
Connection.connect and Connection.disconnect both needed to be overridden by rendered_connection, and there's probably some additional work to do in respawnShadow_.
The code paths for users dragging blocks around, calling APIs directly, and undoing/redoing events are all slightly different and need testing.

**Comment by RoboErikG** [comment]
The fix is now merged, but we're leaving it in develop until the next release in case there are unexpected side effects.

**Comment by RoboErikG** [comment]
And there were unexpected side effects so this ended up being reverted.

**Comment by BeksOmega** [comment]
Fixed by #3806
